dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/helpers/HelperTypes.java

```diff
@@ -0,0 +1,157 @@
+package com.dat3m.dartagnan.parsers.program.visitors.spirv.helpers;
+
+import com.dat3m.dartagnan.exception.ParsingException;
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.Type;
+import com.dat3m.dartagnan.expression.integers.IntLiteral;
+import com.dat3m.dartagnan.expression.type.AggregateType;
+import com.dat3m.dartagnan.expression.type.ArrayType;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
+
+import java.util.List;
+
+import static com.dat3m.dartagnan.expression.integers.IntBinaryOp.ADD;
+import static com.dat3m.dartagnan.expression.integers.IntBinaryOp.MUL;
+
+public class HelperTypes {
+
+    private static final TypeFactory types = TypeFactory.getInstance();
+    private static final ExpressionFactory expressions = ExpressionFactory.getInstance();
+    private static final IntegerType archType = types.getArchType();
+
+    private HelperTypes() {
+    }
+
+    public static Type getMemberType(String id, Type type, List<Integer> indexes) {
+        if (!indexes.isEmpty()) {
+            id += "[" + indexes.get(0) + "]";
+            if (type instanceof ArrayType aType) {
+                return getArrayMemberType(id, aType, indexes);
+            }
+            if (type instanceof AggregateType aType) {
+                return getStructMemberType(id, aType, indexes);
+            }
+            throw new ParsingException(indexTooDeepError(id));
+        }
+        return type;
+    }
+
+    public static int getMemberOffset(String id, int offset, Type type, List<Integer> indexes) {
+        if (!indexes.isEmpty()) {
+            id += "[" + indexes.get(0) + "]";
+            if (type instanceof ArrayType aType) {
+                return getArrayMemberOffset(id, offset, aType, indexes);
+            }
+            if (type instanceof AggregateType aType) {
+                return getStructMemberOffset(id, offset, aType, indexes);
+            }
+            throw new ParsingException(indexTooDeepError(id));
+        }
+        return offset;
+    }
+
+    public static Expression getMemberAddress(String id, Expression base, Type type, List<Expression> indexes) {
+        if (!indexes.isEmpty()) {
+            id += "[" + indexes.get(0) + "]";
+            if (type instanceof ArrayType aType) {
+                return getArrayMemberAddress(id, base, aType, indexes);
+            }
+            if (type instanceof AggregateType aType) {
+                return getStructMemberAddress(id, base, aType, indexes);
+            }
+            throw new ParsingException(indexTooDeepError(id));
+        }
+        return base;
+    }
+
+    private static Type getArrayMemberType(String id, ArrayType type, List<Integer> indexes) {
+        int index = indexes.get(0);
+        if (type.getNumElements() < 0 || index < type.getNumElements()) {
```

Comment:
Unknown rather than dynamic size. But this is an implementation detail so there is a function `ArrayType.hasKnownNumElements()` (which returns false for negative sizes).

Reply:
Then lets use `type.hasKnownNumElements()` instead of `type.getNumElements() < 0`

Reply:
ok